src/main/java/com/intuit/graphql/orchestrator/visitors/queryVisitors/DeferDirectiveQueryVisitor.java

```diff
@@ -0,0 +1,303 @@
+    private final PruneChildDeferSelectionsModifier PRUNE_CHILD_MODIFIER = PruneChildDeferSelectionsModifier
+            .builder()
+            .nestedDefersAllowed(true)
+            .build();
+
+
+    /*
+    * Functions:
+    *    Updates the field if it contains defer directive.
+    *    If it is a valid deferred node, generate new EI
+    */
+    @Override
+    public TraversalControl visitField(Field node, TraverserContext<Node> context) {
+        if(containsValidDeferDirective(node)) {
+            Node prunedNode = pruneDeferDirective(node);
+
+            if(isValidDeferredSelection(node)) {
+                generatedEIs.add(generateDeferredEI(prunedNode, context));
+            }
+
+            return changeNode(context, prunedNode);
+        }
+
+        return TraversalControl.CONTINUE;
+    }
+
+    /*
+     * Functions:
+     *    Updates the FragmentSpread if it contains defer directive.
+     *    If it is a valid deferred node, generate new EI
+     */
+    @Override
+    public TraversalControl visitFragmentSpread(FragmentSpread node, TraverserContext<Node> context) {
+        if(containsValidDeferDirective(node)) {
+            Node prunedNode = pruneDeferDirective(node);
+
+            if(isValidDeferredSelection(node)) {
+                neededFragmentSpreads.add(node.getName());
+                generatedEIs.add(generateDeferredEI(prunedNode, context));
+            }
+
+            return changeNode(context, prunedNode);
+        }
+
+        return TraversalControl.CONTINUE;
+    }
+
+    @Override
+    public TraversalControl visitInlineFragment(InlineFragment node, TraverserContext<Node> context) {
+        if(containsValidDeferDirective(node)) {
+            Node prunedNode = pruneDeferDirective(node);
+
+            if(isValidDeferredSelection(node)) {
+                generatedEIs.add(generateDeferredEI(prunedNode, context));
+            }
+
+            return changeNode(context, prunedNode);
+        }
+
+        return TraversalControl.CONTINUE;
+    }
+
+    /*
+    * Checks if it is necessary to create ei for deferred field.
+    * Currently, selection should be skipped if all the children field are deferred resulting in an empty selection set.
+    */
+    private boolean isValidDeferredSelection(Node selection) {
+        Optional<SelectionSet> possibleSelectionSet = ((List<Node>)selection.getChildren())
+                .stream()
+                .filter(SelectionSet.class::isInstance)
+                .map(SelectionSet.class::cast)
+                .findAny();
+
+                //it is leaf node with no children
+        return !possibleSelectionSet.isPresent() ||
+                // at least one of the children are not deferred
+                possibleSelectionSet.get().getChildren().stream().anyMatch(child -> !containsValidDeferDirective(child));
+    }
+
+    /*
+    * Verifies that Node has defer directive that is not disabled
+    */
+    private boolean containsValidDeferDirective(Node node) {
```

Comment:
Should we rename to containsEnabledDeferDirective?  I think a disabled defer is also valid directive.